crates/runtime/src/instance.rs

```diff
@@ -320,72 +187,70 @@ pub(crate) struct Instance {
 #[allow(clippy::cast_ptr_alignment)]
 impl Instance {
     /// Return the indexed `VMSharedSignatureIndex`.
-    #[allow(dead_code)]
     fn signature_id(&self, index: SignatureIndex) -> VMSharedSignatureIndex {
-        signature_id(&self.vmctx, &self.offsets, index)
+        unsafe { *self.signature_ids_ptr().add(index.as_u32() as usize) }
     }
 
     /// Return a pointer to the `VMSharedSignatureIndex`s.
-    fn signature_ids_ptr(&mut self) -> *mut VMSharedSignatureIndex {
+    fn signature_ids_ptr(&self) -> *mut VMSharedSignatureIndex {
         unsafe {
-            (&mut self.vmctx as *mut VMContext as *mut u8)
+            (self.vmctx_ptr() as *mut u8)
                 .add(usize::try_from(self.offsets.vmctx_signature_ids_begin()).unwrap())
```

Comment:
We do `(self.vmctx_ptr() as *mut u8).add(...)` enough that it might be nice to factor it into a helper function, like `self.vmctx_plus_offset(...)` or so.